sopel/coretasks.py

```diff
@@ -517,6 +516,11 @@ def _parse_modes(bot, args):
         "Y": module.OPER,
     }
 
+    if clear:
+        modes = {}
+    else:
+        modes = channel.modes
```

Comment:
Is there some rationale for not directly resetting the channel's attribute?

```suggestion
    if clear:
        channel.modes = {}
```

_All_ of the changes below are unnecessary if this functions simply continues operating against `channel.modes`, as it has been. The callable is unthreaded, so no need to worry about races.

Reply:
I know less about threading than async, but couldn't a threaded callable started just before the RPL_CHANNELMODEIS unthreaded one end up reading modes[] while it's not yet rebuilt?

Reply:
That could be solved by making it a `SopelMemory` for the mutex behavior, instead of a plain dict. In fact, it's kind of insane that `channel.modes` isn't already a thread-safe type.

Reply:
As a general principle using `SopelMemory` would probably be good, but do you mean to manually take the lock on it while rebuilding? That would work, but I don't see the advantage, and then I'd need `if clear: [take/release lock]` on both ends of the loop.
If I don't manually take the lock: `modes={a,b}`, we `MODE #chan`, I replace the `SopelMemory` and begin rebuilding it. I `__additem__` `a` which takes and releases the lock, another thread checks `__contains__` `b` which takes and releases the lock, then I continue to add `b`, which lets the other callable think it was is absent when it never was.

Reply:
```
modes = {}
if not clear:
    modes.update(channel.modes)

# replace channel.modes by modes
# then:
channel.modes = modes
```

Then you won't have the problem of race condition while updating `channel.modes`. The main problem here is that Sopel uses both async to deal with the socket and multi-threading to deal with command handling. And no, you can't complain as much as I do on that topic, no need to try.

Reply:
Thanks!